src/SAWScript/Proof.hs

```diff
@@ -753,156 +1340,195 @@ checkEvidence sc db = \e p -> do hyps <- Map.keysSet <$> readIORef (theoremMap d
-          Nothing -> fail $ unlines ["Assume evidence expected function prop", showTerm ptm]
-          Just (_lnm, ty, body) ->
-            do ok <- scConvertible sc False ty p'
-               unless ok $ fail $ unlines
-                   [ "Assume evidence types do not match"
-                   , showTerm ty
-                   , showTerm p'
-                   ]
-               unless (looseVars body == emptyBitSet) $ fail $ unlines
-                   [ "Assume evidence cannot be used on a dependent proposition"
-                   , showTerm ptm
-                   ]
-               (d,sy) <- check (Set.insert n hyps) e' (Prop body)
-               return (Set.delete n d, sy)
-
-      ForallEvidence x e' ->
-        case asPi ptm of
-          Nothing -> fail $ unlines ["Assume evidence expected function prop", showTerm ptm]
-          Just (_lnm, ty, body) ->
-            do let ty' = ecType x
-               ok <- scConvertible sc False ty ty'
-               unless ok $ fail $ unlines
-                 ["Forall evidence types do not match"
-                 , showTerm ty'
-                 , showTerm ty
-                 ]
-               x' <- scExtCns sc x
-               body' <- instantiateVar sc 0 x' body
-               check hyps e' (Prop body')
+        do sqt' <- traverseSequentWithFocus (evalProp sc vars) sqt
+           check nenv e' sqt'
+
+      ConversionEvidence sqt' e' ->
+        do ok <- convertibleSequents sc sqt sqt'
+           unless ok $ fail $ unlines
+             [ "Converted sequent does not match goal"
+             , prettySequent defaultPPOpts nenv sqt
+             , prettySequent defaultPPOpts nenv sqt'
+             ]
+           check nenv e' sqt'
+
+      NormalizeSequentEvidence sqt' e' ->
+        do ok <- normalizeSequentSubsumes sc sqt' sqt
+           unless ok $ fail $ unlines
+             [ "Normalized sequent does not subsume goal"
+             , prettySequent defaultPPOpts nenv sqt
+             , prettySequent defaultPPOpts nenv sqt'
+             ]
+           check nenv e' sqt'
+
+      StructuralEvidence sqt' e' ->
+        do ok <- sequentSubsumes sc sqt' sqt
+           unless ok $ fail $ unlines
+             [ "Sequent does not subsume goal"
+             , prettySequent defaultPPOpts nenv sqt
+             , prettySequent defaultPPOpts nenv sqt'
+             ]
+           check nenv e' sqt'
+
+      AxiomEvidence ->
+        do ok <- sequentIsAxiom sc sqt
+           unless ok $ fail $ unlines
+             [ "Sequent is not an instance of the sequent calculus axiom"
+             , prettySequent defaultPPOpts nenv sqt
+             ]
+           return (mempty, ProvedTheorem mempty)
+
+      CutEvidence p ehyp egl ->
+        do d1 <- check nenv ehyp (addHypothesis p sqt)
+           d2 <- check nenv egl  (addNewFocusedGoal p sqt)
+           return (d1 <> d2)
+
+      IntroEvidence x e' ->
+        -- TODO! Check that the given ExtCns is fresh for the sequent
```

Comment:
Are you planning to do this, or should this warrant an issue as a reminder to do later?

Reply:
I'll open an issue about this. I'm not 100% sure what is the best way to hande this.